services/video/client.go

```diff
@@ -0,0 +1,110 @@
+package video
+
+import (
+	"context"
+	"errors"
+	"io"
+	"time"
+
+	commonpb "go.viam.com/api/common/v1"
+	pb "go.viam.com/api/service/video/v1"
+	"go.viam.com/utils/protoutils"
+	"go.viam.com/utils/rpc"
+	"google.golang.org/protobuf/types/known/timestamppb"
+
+	"go.viam.com/rdk/logging"
+	"go.viam.com/rdk/resource"
+)
+
+// client is a video service client that talks to a gRPC video service.
+type client struct {
+	resource.Named
+	resource.TriviallyReconfigurable
+	resource.TriviallyCloseable
+	name   string
+	client pb.VideoServiceClient
+	logger logging.Logger
+}
+
+// NewClientFromConn creates a new video service client from a gRPC connection.
+func NewClientFromConn(
+	ctx context.Context,
+	conn rpc.ClientConn,
+	remoteName string,
+	name resource.Name,
+	logger logging.Logger,
+) (Service, error) {
+	c := pb.NewVideoServiceClient(conn)
+	return &client{
+		Named:  name.PrependRemote(remoteName).AsNamed(),
+		name:   name.Name,
+		client: c,
+		logger: logger,
+	}, nil
+}
+
+// GetVideo calls the GetVideo method on the video service.
+// It writes the received video data to the provided writer.
+func (c *client) GetVideo(
+	ctx context.Context,
+	startTime, endTime time.Time,
+	videoCodec, videoContainer, requestID string,
+	extra map[string]interface{},
+	w io.Writer,
+) error {
+	ext, err := protoutils.StructToStructPb(extra)
+	if err != nil {
+		return err
+	}
+	var startTS *timestamppb.Timestamp
+	if !startTime.IsZero() {
+		startTS = timestamppb.New(startTime)
+	}
+	var endTS *timestamppb.Timestamp
+	if !endTime.IsZero() {
+		endTS = timestamppb.New(endTime)
+	}
+	req := &pb.GetVideoRequest{
+		Name:           c.name,
+		StartTimestamp: startTS,
+		EndTimestamp:   endTS,
+		VideoCodec:     videoCodec,
+		VideoContainer: videoContainer,
+		RequestId:      requestID,
+		Extra:          ext,
+	}
+	stream, err := c.client.GetVideo(ctx, req)
+	if err != nil {
+		return err
+	}
+	for {
+		chunk, err := stream.Recv()
+		if errors.Is(err, io.EOF) {
+			break
+		}
+		if err != nil {
+			return err
+		}
+		_, err = w.Write(chunk.GetVideoData())
+		if err != nil {
+			return err
+		}
+	}
```

Comment:
all of our resource method calls have a context with timeout right? Thinking of ways this could possibly hang

Reply:
also can we add tests to cover all branches of this loop?

Reply:
> all of our resource method calls have a context with timeout right? Thinking of ways this could possibly hang

Good call out here, it is quite confusing.

The context gets passed through the stream, `stream.Recv()` will error out when the context cancels, so we do not need an explicit check on ctx.

Adding comment to clarify this